### PR TITLE
Integrate low-level (point nav + social nav + pick + place) policies for the demo

### PR DESCRIPTION
**Pull request by jimmytyyang**
The initial integration of point nav (find the object), pick, social nav (return to human after picking up the object), and place neural network skills are finished.

There are still a few things that need to be polished, such as the place policy drops the object at a few steps without too much arm animation, etc.

### CODE REVIEW

**Comment by jimmytyyang** [comment]
CLI to run the demo

`python examples/siro_sandbox/sandbox_app.py --app-state fetch --disable-inverse-kinematics --never-end --width 1920 --height 1080 --gui-controlled-agent-index 1 --save-gfx-replay-keyframes --save-filepath-base "/Users/jimmytyyang/Downloads/soc_nav_video/video_v8" --ang-speed 10 --cfg examples/siro_sandbox/configs/fetch.yaml --cfg-opts \~habitat.task.measurements.agent_blame_measure habitat.task.actions.agent_0_oracle_nav_action.type=OracleNavCoordAction habitat_baselines.evaluate=True habitat.simulator.step_physics=True`

File: examples/siro_sandbox/controllers/baselines_controller.py
Comment (0mdc) [inline examples/siro_sandbox/controllers/baselines_controller.py:1]: General comment, no need to fix now:

Some code here is Fetch-specific and may be eventually moved to `app_state_fetch.py`.

File: habitat-lab/habitat/config/default_structured_configs.py [hunk 1]
```diff
@@ -268,10 +271,25 @@ class BaseVelocityLegAnimationActionConfig(ActionConfig):
...
+    # The max longitudinal and lateral linear speeds of the robot
+    longitudinal_lin_speed: float = 10.0
+    lateral_lin_speed: float = 10.0
     lin_speed: float = 10.0
+    # The max angular speed of the robot
     ang_speed: float = 10.0
-    allow_dyn_slide: bool = True
+    # If we want to do sliding or not
+    allow_dyn_slide: bool = False
+    # If the condition of sliding includs the checking of rotation
```
Comment (0mdc) [inline habitat-lab/habitat/config/default_structured_configs.py:282]: Typo: `includes`

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py [hunk 2]
```diff
@@ -691,6 +688,84 @@ def step(self, *args, **kwargs):
...
+        self._load_animation()
+        self._play_i = 0
+        self._play_length_data = len(self._leg_data)
+        self._play_i_perframe = self._config.get("play_i_perframe")
+
+    def _load_animation(self):
+        first_row = True
+        time_i = 0
+        with open(self._checkpoint, newline="") as csvfile:
+            spamreader = csv.reader(csvfile, delimiter=" ", quotechar="|")
```
Comment (0mdc) [inline habitat-lab/habitat/tasks/rearrange/actions/actions.py:708]: `spamreader`? :)

File: habitat-lab/habitat/tasks/rearrange/multi_task/composite_sensors.py [hunk 3]
```diff
@@ -323,3 +324,66 @@ def update_metric(self, *args, task, **kwargs):
...
+    def reset_metric(self, *args, **kwargs):
+        self._visit_loc_round_half = {}
+        self._visit_loc_round_0 = {}
+        self._visit_loc_round_1 = {}
+        self.update_metric(
+            *args,
+            **kwargs,
+        )
+
+    def normal_round(self, n):
```
Comment (0mdc) [inline habitat-lab/habitat/tasks/rearrange/multi_task/composite_sensors.py:358]: Nit: Could be renamed `round_to_nearest()`

File: habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py [hunk 4]
```diff
@@ -1176,3 +1217,344 @@ def update_metric(self, *args, task, **kwargs):
...
+    def _check_human_frame(self, obs):
+        if not self._check_human_in_frame:
+            return True
+
+        panoptic = obs["agent_0_articulated_agent_arm_panoptic"]
+        return (
+            np.sum(panoptic == self._human_id) > self._human_detect_threshold
+        )
+
+    def _check_look_at_human(self, human_pos, robot_pos):
```
Comment (0mdc) [inline habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py:1350]: Nit: Could be renamed `_check_robot_facing_human`

File: habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py [hunk 5]
```diff
@@ -1176,3 +1217,344 @@ def update_metric(self, *args, task, **kwargs):
...
+                robot_to_human_min_step = self._step
+            else:
+                robot_to_human_min_step = float("inf")
+
+            # Update the minimum SPL
+            self._min_start_end_episode_step = min(
+                self._min_start_end_episode_step, robot_to_human_min_step
+            )
+
+        # Compute the SPL before finding the human
```
Comment (0mdc) [inline habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py:1474]: Pardon my ignorance - what is `SPL` in this context?

**Review by 0mdc — APPROVED** [review]
LGTM. I added a few minor comments.

I have little knowledge of this domain, I encourage you to wait for @xavierpuigf's input before merging.

File: examples/siro_sandbox/controllers/baselines_controller.py [hunk 6]
```diff
@@ -291,19 +293,68 @@ def get_cartesian_obj_coords(self, obj_trans, env):
 
         return rho, phi
 
+    def get_geodesic_distance_obj_coords(self, obj_trans, env):
```
Comment (rpartsey) [inline examples/siro_sandbox/controllers/baselines_controller.py:296]: `env` argument is not used in this function, can be removed

File: examples/siro_sandbox/controllers/baselines_controller.py [hunk 7]
```diff
@@ -291,19 +293,68 @@ def get_cartesian_obj_coords(self, obj_trans, env):
...
+        # TODO: we use geodesic_distance here
+        articulated_agent_T = self.get_articulated_agent().base_transformation
+        rel_pos = articulated_agent_T.inverted().transform_point(obj_trans)
+        rho, phi = cartesian_to_polar(rel_pos[0], rel_pos[1])
```
Comment (rpartsey) [inline examples/siro_sandbox/controllers/baselines_controller.py:300]: This code duplicates `get_cartesian_obj_coords` function. Can we call `get_cartesian_obj_coords` to get `rho` and `phi` instead?

**Review by rpartsey — APPROVED** [review]

**Comment by jimmytyyang** [comment]
The Google folder link to download neural network skills (point-nav, pick, social-nav, place): https://drive.google.com/drive/folders/1Ddmt30hBUPL0pwfsXWj-K4CQYgOGsZNg?usp=sharing

File: habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py (hunk 5 above)
Comment (jimmytyyang) [reply habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py:1474]: Sure, no problem! 

SPL is the metric that measures the performance of how quickly the robot can find the humanoid. The higher the SPL value is, the better, and it is bounded to be in the range of 0 and 1. 

You can read the paper section 4.1 to get how exactly we compute it

File: examples/siro_sandbox/configs/fetch.yaml [hunk 8]
```diff
@@ -66,7 +74,15 @@ habitat:
         type: OracleNavCoordAction
 
       agent_0_base_velocity:
-        lin_speed: 6.0
+        allow_dyn_slide: True
+        collision_threshold: 1e-5
+        navmesh_offset: [[0.0, 0.0], [0.275, 0.0]]
+        enable_lateral_move: False
+        longitudinal_lin_speed: 10.0
+        lateral_lin_speed: 10.0
```
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply examples/siro_sandbox/configs/fetch.yaml:82]: Hi Xavi! Thanks for this. During training, the point nav max speed is 10, let me change it to 5 to see if that will affect the performance.

File: examples/siro_sandbox/controllers/gui_controller.py [hunk 9]
```diff
@@ -458,17 +458,18 @@ def act(self, obs, env):
         fixup = filtered_query_pos - target_query_pos
         self._humanoid_controller.obj_transform_base.translation += fixup
 
-        if self._hint_reach_pos:
-            self._humanoid_controller.calculate_reach_pose(
-                self._hint_reach_pos
-            )
-            self._hint_reach_pos = None
+        # TODO: remove the joint angles overwrite here
```
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply examples/siro_sandbox/controllers/gui_controller.py:461]: Hi Xavi, it seems that the human walking animation is affected here. Could you please check that for me? Thank you!

File: examples/siro_sandbox/configs/fetch.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (xavierpuigf) [reply examples/siro_sandbox/configs/fetch.yaml:82]: It is fine if during training we do 10! No need to retrain :) Just wanted to clarify that this was not used during eval.
If so (small suggestion), maybe you could add a comment on top of the confiig saying that this config is to **train** skills.

**Comment by jimmytyyang** [comment]
The New CLI without `OracleNavCoordAction`
`python examples/siro_sandbox/sandbox_app.py --app-state fetch --disable-inverse-kinematics --never-end --width 1920 --height 1080 --gui-controlled-agent-index 1 --save-gfx-replay-keyframes --save-filepath-base "/Users/jimmytyyang/Downloads/soc_nav_video/video_v8" --ang-speed 10 --cfg examples/siro_sandbox/configs/fetch.yaml --cfg-opts \~habitat.task.measurements.agent_blame_measure  habitat_baselines.evaluate=True habitat.simulator.step_physics=True`

**Comment by jimmytyyang** [comment]
Note that the point nav policy was updated. Please go to Google link above and download the new model (checkpoint)
